### DISCUSSION TITLE
Rendering doesn't wait for serverPrefetch with Pinia store

### DISCUSSION THREAD

**Discussion by DannyZB** [vikejs/vike#548] [Discussion] [answered]
### Description

I am having an issue with rendering on SSR doesn't wait for onServerPrefetch to be done before returning from renderToNodeStream
Ends up with wrong data rendered and in the Pinia store..
```

async function onBeforeRender(pageContext)
{
    console.log('onBeforeRender started')
    if(typeof document !== 'undefined') { // client side
        return {
            pageContext: {
            }
        }
    }

    console.log('onBeforeRender: App creation started')
    const { app, store } = createApp(pageContext)
    console.log('onBeforeRender: App creation done')
    const stream = renderToNodeStream(app)
    console.log('onBeforeRender: Stream created')

    // const initialStoreState = toRawStoreValue(store.state.value)
    const initialStoreState = store.state.value
    console.log('onBeforeRender: Initial Store State Set')

    return {
        pageContext: {
            initialStoreState,
            stream
        },
    }
}


async function render(pageContext) {
    const { stream } = pageContext
    console.log('Document rendering started')

    const documentHtml = escapeInject`<!DOCTYPE html>
  <html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <link rel="icon" href="/favicon.ico">
    <title></title>
  </head>
  <body>
    <div id="app">${stream}</div>
  </body>
  </html>`

    return {
        documentHtml,
        pageContext: {
            enableEagerStreaming: true
        }
    }
}
 
```
onBeforeRender: Initial Store State Set is logged before serverPrefetch is fired.
On the surface I did exactly like in the Pinia example. Not sure what's broken.

**Comment by brillout** [comment]
The thing here is that `onServerPrefetch()` is called *during* the stream. This means that when you access the store before passing the stream to VPS, then `onServerPrefetch()` wasn't called yet.

To fix this you can return a `pageContext` promise, see https://vite-plugin-ssr.com/stream#initial-data-after-streaming.

**Comment by DannyZB** [comment]
Tried it, didn't help.
Also, I followed the example for vue-pinia https://github.com/brillout/vite-plugin-ssr/blob/main/examples/vue-pinia/renderer/_default.page.server.js

You don't use that there.

I tested with that and the prefetch is done before returning the response to the client, but after the pageContextPromise is called.

**Reply by brillout** [reply]
You can now return a `pageContext` promise function, see the updated docs https://vite-plugin-ssr.com/stream#initial-data-after-streaming. Released `v0.4.57`. That should do the trick.

**Comment by DannyZB** [comment]
In my code it looks like the serverPrefetch hook just doesn't get waited for.
Removed Pinia and replaced with await sleep(1000) and the same issue is there.

**Reply by brillout** [reply]
The thing here is that addition `pageContext` needs to be provided *after* the stream ends. You can now do that with the latest release, see my comment above.

**Comment by DannyZB** [comment, marked as answer]
The latest release resolved the issue.
Thank you!
